agent/consul/state/catalog.go

```diff
@@ -4551,7 +4555,7 @@ func (s *Store) ServiceTopology(
 		if downstream.Service.Kind == structs.ServiceKindConnectProxy {
 			sn = structs.NewServiceName(downstream.Service.Proxy.DestinationServiceName, &downstream.Service.EnterpriseMeta)
 		}
-		if _, ok := tproxyMap[sn]; !ok && !downstream.Service.Connect.Native && downstreamSources[sn.String()] != structs.TopologySourceRegistration {
+		if _, ok := proxyMap[sn]; !ok && downstreamSources[sn.String()] != structs.TopologySourceRegistration {
 			// If downstream is not a transparent proxy or connect native, remove references
 			delete(downstreamSources, sn.String())
```

Comment:
Thanks for the review @kisunji, I have pushed the requests changes. One thing I am not 100% confident is this if here. Once we remove from `downStreamSources` the if in the line above probably cannot find `structs.TopologySourceRegistration` anymore (even if it was in there before). So I wonder if we should delay the deletion or can this occurrence not happen?

Reply:
Sorry I don't quite understand your concern here. I don't think anything about this if conditional is being changed:

Before: `If downstream is NOT TProxy && NOT Connect && NOT structs.TopologySourceRegistration`

After: `If downstream is NOT (TProxy OR Connect) && NOT structs.TopologySourceRegistration`

Am I misunderstanding your question?

Reply:
I was mainly questioning if it is already wrong before my changes :) Assuming we have duplicated a `sn` (which afaik is still possible, since the dedup happens on the node & service combo) then after the first `if` this `sn` would have been removed from `downstreamSources` and so it will never trigger the `if` (and therefore the `continue` inside) again. My code does not make this any/much worse (I guess?), but I don't understand the code completely either.

Reply:
Ah, the `sn` (ServiceName) is the logical name for the service (e.g. `cproxy`).

Our problem was duplicated CheckServiceNodes (service instances).

`cproxy` ServiceName had `['cproxy-https', 'cproxy-http', 'cproxy-https']` CheckServiceNodes.

So no worries about duplicated service names here!